master/internal/db/postgres_trial.go

```diff
@@ -441,3 +450,288 @@ WHERE t.id = $1;
+	apiv1.OrderBy_ORDER_BY_UNSPECIFIED: SortDirectionAsc,
+	apiv1.OrderBy_ORDER_BY_ASC:         SortDirectionAsc,
+	apiv1.OrderBy_ORDER_BY_DESC:        SortDirectionDescNullsLast,
+}
+
+// This allows dot on top of whats allowed in existing regex validField.
+var safeString = regexp.MustCompile(`^[a-zA-Z0-9_\.\-]+$`)
+
+func hParamAccessor(hp string) string {
+	nesting := strings.Split(hp, ".")
+	nestingWithQuotes := []string{}
+	for _, n := range nesting {
+		nestingWithQuotes = append(nestingWithQuotes, fmt.Sprintf("'%s'", n))
+	}
+	return "hparams->>" + strings.Join(nestingWithQuotes, "->>")
+}
+
+// ApplyTrialPatch applies a patch operation to a set of Trials.
+func (db *PgDB) ApplyTrialPatch(q *bun.UpdateQuery, 
+	payload *apiv1.TrialPatch) (*bun.UpdateQuery, error) {
+	// takes an update query and adds the Set clauses for the patch
+
+	if len(payload.AddTag) > 0 || len(payload.RemoveTag) > 0 {
+
+		addTags := map[string]string{}
+		for _, tag := range payload.AddTag {
+			addTags[tag.Key] = ""
+		}
+
+		removeTags := []string{}
+		for _, tag := range payload.RemoveTag {
+			removeTags = append(removeTags, tag.Key)
+		}
+
+		q = q.Set("tags = tags || ? - ?::text[]", addTags, pgdialect.Array(removeTags))
+	}
+	return q, nil
+}
+
+// TrialsColumnForNamespace returns the correct namespace for a TrialSorter
+func (db *PgDB) TrialsColumnForNamespace(namespace apiv1.TrialSorter_Namespace, 
+	field string) (string, error) {
+	if !safeString.MatchString(field) {
+		return "", fmt.Errorf("%s filter %s contains possible SQL injection", namespace, field)
+	}
+	switch namespace {
+	case apiv1.TrialSorter_NAMESPACE_TRIALS_UNSPECIFIED:
+		return field, nil
+	case apiv1.TrialSorter_NAMESPACE_HPARAMS:
+		return hParamAccessor(field), nil
+	case apiv1.TrialSorter_NAMESPACE_TRAINING_METRICS:
+		return fmt.Sprintf("training_metrics->>'%s'", field), nil
+	case apiv1.TrialSorter_NAMESPACE_VALIDATION_METRICS:
+		return fmt.Sprintf("validation_metrics->>'%s'", field), nil
+	default:
+		return field, nil
+	}
+}
+
+func conditionalForNumberRange(min *wrappers.DoubleValue, max *wrappers.DoubleValue) string {
+	switch {
+	case min != nil && max != nil:
+		return fmt.Sprintf("BETWEEN %f AND %f", min.Value, max.Value)
+	case min != nil :
+		return fmt.Sprintf(" > %f", min.Value)
+	case max != nil :
+		return fmt.Sprintf(" < %f", max.Value)
+	default: 
+	return notNull
+	}
+}
+
+func conditionalForDateTimeRange(q *bun.SelectQuery, column string, dateTime *apiv1.TimeRangeFilter) {
+	startTime := dateTime.IntervalStart
+	endTime := dateTime.IntervalEnd
+	switch {
+	case startTime != nil && endTime != nil:
+		q.Where("? BETWEEN ? AND ?", column, startTime.AsTime(), endTime.AsTime())
+	case startTime != nil:
+		q.Where("? > ?", column, startTime.AsTime())
+	case endTime != nil:
+		q.Where("? < ?", column, endTime.AsTime())
+	}
+}
+
+// FilterTrials queries for Trials matching the supplied TrialFilters.
+func (db *PgDB) FilterTrials(q *bun.SelectQuery, 
```

Comment:
I think for readability it may be useful to distinguish database funcs which build and run a query and return the result vs the ones which just build a query for use elsewhere. I am not sure we already have a precedents for this case. so perhaps rename this function to one of:
- `BuildFilterTrialsQuery`
- `MakeFilterTrialsQuery`
- `BuildFilterTrials`
- `MakeFilterTrials`
- `FilterTrialsQuery`
or similar

Reply:
just pick one for me lol 🙃

Reply:
in seriousness though, im partial to `BuildTrialFiltersQuery` (not on the list but much in keeping)